### PR TITLE
[57812] Added model and migration for custom field of type hierarchy

### PR DESCRIPTION
**Pull request by apfohl**
# Ticket
https://community.openproject.org/projects/openproject/work_packages/57812/activity

# What are you trying to accomplish?
This PR implements a new data structure for storing hierarchical item for use in the new custom field of type hierarchy. Therefore every custom field of type `hierarchy` now holds a reference to a ´root´-item. From there the hierarchy starts.

# What approach did you choose and why?
I used the ´closure_tree´ gem as discussed in the architecture decision.

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: app/models/custom_value/hierarchy_strategy.rb
Comment (apfohl) [inline app/models/custom_value/hierarchy_strategy.rb:1]: 🟡 I'm not sure about this change, if it already goes to deep into frontend.

File: config/initializers/custom_field_format.rb
Comment (apfohl) [inline config/initializers/custom_field_format.rb:1]: 🟡 I'm not sure about this change, if it already goes to deep into frontend.

File: lib/api/v3/utilities/custom_field_injector.rb
Comment (apfohl) [inline lib/api/v3/utilities/custom_field_injector.rb:1]: 🟡 I'm not sure about this change, if it already goes to deep into frontend.

File: app/models/custom_field/Hierarchy/item.rb
Comment (Kharonus) [inline app/models/custom_field/Hierarchy/item.rb:1]: 🟡 you might want to rename the directory to `snake_case`, so in this time lowering the case of the first letter ;)

File: app/models/custom_value/hierarchy_strategy.rb
Comment (Kharonus) [reply app/models/custom_value/hierarchy_strategy.rb:1]: I'm not sure, what I read here at all. 😁 Where an when is this strategy called?

File: config/initializers/custom_field_format.rb
Comment (Kharonus) [reply config/initializers/custom_field_format.rb:1]: 🔴 I'd either not commit that or already put that behind a feature flag, as we do not want to show hierarchy in the creation dropdown, until we are done with it.

I'd tend to adding the feature flag directly. If you want, I can commit that to your branch

File: lib/api/v3/utilities/custom_field_injector.rb
Comment (Kharonus) [reply lib/api/v3/utilities/custom_field_injector.rb:1]: 🟢  no harm done here, so fine for me.

**Review by Kharonus — CHANGES_REQUESTED** [review]

File: config/initializers/custom_field_format.rb
Comment (Kharonus) [reply config/initializers/custom_field_format.rb:1]: solved

Comment (Kharonus) [reply config/initializers/custom_field_format.rb:1]: solved

**Review by Kharonus — APPROVED** [review]
LGTM

**Pull request merged**